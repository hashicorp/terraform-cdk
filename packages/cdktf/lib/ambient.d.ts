// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
declare module 'archiver';