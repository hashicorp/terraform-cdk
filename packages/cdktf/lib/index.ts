// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { validateEnvironment } from "./validateEnvironment";
validateEnvironment();

export * from "./tokens";
export * from "./terraform-stack";
export * from "./terraform-element";
export * from "./terraform-module";
export * from "./terraform-provider";
export * from "./terraform-resource";
export * from "./terraform-data-source";
export * from "./terraform-output";
export * from "./complex-computed-list";
export * from "./resource";
export * from "./testing";
export * as testingMatchers from "./testing/matchers";
export * from "./app";
export * from "./terraform-backend";
export * from "./backends";
export * from "./terraform-remote-state";
export * from "./terraform-local";
export * from "./terraform-variable";
export * from "./terraform-hcl-module";
export * from "./runtime";
export * from "./terraform-dependable";
export * from "./manifest";
export * from "./terraform-asset";
export * from "./synthesize";
export * from "./annotations";
export * from "./aspect";
export * from "./terraform-functions";
export * from "./terraform-operators";
export * from "./tfExpression";
export * from "./terraform-addressable";
export * from "./terraform-iterator";
export * from "./terraform-provisioner";
export * from "./terraform-conditions";
export * from "./terraform-count";

// required for JSII because Fn extends from it
export * from "./functions/terraform-functions.generated";
