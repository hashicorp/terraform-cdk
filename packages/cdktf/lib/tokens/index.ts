// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
export * from "./lazy";
export * from "./resolvable";
export * from "./string-fragments";
export * from "./token";
