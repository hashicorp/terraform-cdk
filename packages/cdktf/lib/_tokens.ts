// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, DefaultTokenResolver, StringConcat } from "./tokens";
import { Construct } from "constructs";

const TOKEN_RESOLVER = new DefaultTokenResolver(new StringConcat());

// eslint-disable-next-line jsdoc/require-jsdoc
export function resolve<T>(scope: Construct, obj: T, preparing = false): T {
  return Tokenization.resolve(obj, {
    scope,
    preparing,
    resolver: TOKEN_RESOLVER,
  });
}
