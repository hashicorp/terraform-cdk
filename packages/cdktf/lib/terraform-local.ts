// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";
import { Token } from "./tokens";
import { ref } from "./tfExpression";
import { IResolvable } from "./tokens/resolvable";
import { ITerraformAddressable } from "./terraform-addressable";

// eslint-disable-next-line jsdoc/require-jsdoc
export class TerraformLocal
  extends TerraformElement
  implements ITerraformAddressable
{
  private _expression: any;

  constructor(scope: Construct, id: string, expression: any) {
    super(scope, id, "local");

    this._expression = expression;
  }

  public set expression(value: any) {
    this._expression = value;
  }

  public get expression() {
    return Token.asAny(this.interpolation());
  }

  public get asString(): string {
    return Token.asString(this.interpolation());
  }

  public get asNumber(): number {
    return Token.asNumber(this.interpolation());
  }

  public get asList(): string[] {
    return Token.asList(this.interpolation());
  }

  public get asBoolean(): IResolvable {
    return this.interpolation();
  }

  private interpolation(): any {
    return ref(`local.${this.friendlyUniqueId}`, this.cdktfStack);
  }

  public toTerraform(): any {
    return {
      locals: {
        [this.friendlyUniqueId]: this._expression,
      },
    };
  }
  public toMetadata(): any {
    if (!Object.keys(this.rawOverrides).length) {
      return {};
    }

    return {
      overrides: {
        local: Object.keys(this.rawOverrides),
      },
    };
  }
}
