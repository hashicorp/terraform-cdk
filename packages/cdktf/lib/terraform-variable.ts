// Copyright (c) HashiCorp, Inc
// Licensed under the Mozilla Public License 2.0
import { Construct } from "constructs";
import { TerraformElement } from "./terraform-element";
import { keysToSnakeCase, deepMerge } from "./util";
import { Token } from "./tokens";
import { Expression, ref } from "./tfExpression";
import { IResolvable } from "./tokens/resolvable";
import { ITerraformAddressable } from "./terraform-addressable";

// eslint-disable-next-line jsdoc/require-jsdoc
export abstract class VariableType {
  public static readonly STRING = "string";
  public static readonly NUMBER = "number";
  public static readonly BOOL = "bool";
  public static readonly ANY = "any";

  public static readonly LIST = "list";
  public static readonly MAP = "map";
  public static readonly SET = "set";

  public static readonly LIST_STRING = "list(string)";
  public static readonly LIST_NUMBER = "list(number)";
  public static readonly LIST_BOOL = "list(bool)";

  public static readonly MAP_STRING = "map(string)";
  public static readonly MAP_NUMBER = "map(number)";
  public static readonly MAP_BOOL = "map(bool)";

  public static readonly SET_STRING = "set(string)";
  public static readonly SET_NUMBER = "set(number)";
  public static readonly SET_BOOL = "set(bool)";

  public static list(type: string) {
    return `list(${type})`;
  }

  public static map(type: string) {
    return `map(${type})`;
  }

  public static set(type: string) {
    return `set(${type})`;
  }

  public static tuple(...elements: string[]) {
    return `tuple(${elements.join(", ")})`;
  }

  public static object(attributes: { [key: string]: string }) {
    return `object({${Object.keys(attributes)
      .map((k) => k + "=" + attributes[k])
      .join(", ")}})`;
  }
}

export interface TerraformVariableValidationConfig {
  readonly condition: Expression;
  readonly errorMessage: string;
}

export interface TerraformVariableConfig {
  readonly default?: any;
  readonly description?: string;

  /**
   * The type argument in a variable block allows you to restrict the type of value that will be accepted as the value for a variable. If no type constraint is set then a value of any type is accepted.
   *
   * While type constraints are optional, we recommend specifying them; they serve as easy reminders for users of the module, and allow Terraform to return a helpful error message if the wrong type is used.
   *
   * Type constraints are created from a mixture of type keywords and type constructors. The supported type keywords are:
   *
   * - string
   * - number
   * - bool
   *
   * The type constructors allow you to specify complex types such as collections:
   *
   * - list(\<TYPE\>)
   * - set(\<TYPE\>)
   * - map(\<TYPE\>)
   * - object({\<ATTR NAME\> = \<TYPE\>, ... })
   * - tuple([\<TYPE\>, ...])
   *
   * The keyword any may be used to indicate that any type is acceptable. For more information on the meaning and behavior of these different types, as well as detailed information about automatic conversion of complex types, see {@link https://www.terraform.io/docs/configuration/types.html|Type Constraints}.
   *
   * If both the type and default arguments are specified, the given default value must be convertible to the specified type.
   */
  readonly type?: string;

  readonly sensitive?: boolean;

  /*
   * The nullable argument in a variable block controls whether the module caller may assign the value null to the variable.
   */
  readonly nullable?: boolean;

  /**
   * Specify arbitrary custom validation rules for a particular variable using a validation block nested within the corresponding variable block
   */
  readonly validation?: TerraformVariableValidationConfig[];
}

// eslint-disable-next-line jsdoc/require-jsdoc
export class TerraformVariable
  extends TerraformElement
  implements ITerraformAddressable
{
  public readonly default?: any;
  public readonly description?: string;
  public readonly type?: string;
  public readonly sensitive?: boolean;
  public readonly nullable?: boolean;
  private _validation?: TerraformVariableValidationConfig[];

  constructor(scope: Construct, id: string, config: TerraformVariableConfig) {
    super(scope, id, "var");

    this.default = config.default;
    this.description = config.description;
    this.type = config.type;
    this.sensitive = config.sensitive;
    this.nullable = config.nullable;
    this._validation = config.validation;
  }

  public get stringValue(): string {
    return Token.asString(this.interpolation());
  }

  public get numberValue(): number {
    return Token.asNumber(this.interpolation());
  }

  public get listValue(): string[] {
    return Token.asList(this.interpolation());
  }

  public get booleanValue(): IResolvable {
    return this.interpolation();
  }

  public get value(): any {
    return Token.asAny(this.interpolation());
  }

  public get validation(): TerraformVariableValidationConfig[] | undefined {
    return this._validation;
  }

  public addValidation(validation: TerraformVariableValidationConfig) {
    if (!this._validation) {
      this._validation = [];
    }

    this._validation.push(validation);
  }

  private interpolation(): IResolvable {
    return ref(`var.${this.friendlyUniqueId}`, this.cdktfStack);
  }

  public synthesizeAttributes(): { [key: string]: any } {
    return {
      default: this.default,
      description: this.description,
      type: this.type,
      sensitive: this.sensitive,
      nullable: this.nullable,
      validation: this.validation?.map((validation) => ({
        error_message: validation.errorMessage,
        condition: validation.condition,
      })),
    };
  }

  public toTerraform(): any {
    return {
      variable: {
        [this.friendlyUniqueId]: deepMerge(
          keysToSnakeCase(this.synthesizeAttributes()),
          this.rawOverrides
        ),
      },
    };
  }
}
