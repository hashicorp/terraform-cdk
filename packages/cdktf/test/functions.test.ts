// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Testing, TerraformStack, TerraformOutput, Fn } from "../lib";
import { TerraformVariable } from "../lib/terraform-variable";
import { TerraformLocal } from "../lib/terraform-local";
import { ref } from "../lib/tfExpression";
import { Token } from "../lib/tokens/token";

test("static values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.abs(-42),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${abs(-42)}"
        }
      }
    }"
  `);
});

test("dynamic values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.abs(variable.value),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${abs(var.test-var)}"
        }
      },
      "variable": {
        "test-var": {
          "type": "number"
        }
      }
    }"
  `);
});

test("spreaded mixed values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.max([10, variable.value, 200]),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${max(10, var.test-var, 200)}"
        }
      },
      "variable": {
        "test-var": {
          "type": "number"
        }
      }
    }"
  `);
});

test("spreaded token value", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "list(number)",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.max([variable.value]),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${max(var.test-var)}"
        }
      },
      "variable": {
        "test-var": {
          "type": "list(number)"
        }
      }
    }"
  `);
});

test("string values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.parseint("-210", 10),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${parseint(\\"-210\\", 10)}"
        }
      }
    }"
  `);
});

test("mixed string spreads values", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.format("There are %d out of %d lights are on in %s", [
      variable.value,
      4,
      "Hamburg",
    ]),
  });
  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${format(\\"There are %d out of %d lights are on in %s\\", var.test-var, 4, \\"Hamburg\\")}"
        }
      },
      "variable": {
        "test-var": {
          "type": "number"
        }
      }
    }"
  `);
});

test("combined functions", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const index = new TerraformVariable(stack, "index", {
    type: "number",
  });

  const list = new TerraformVariable(stack, "list", {
    type: `list(object({
      internal = number
      external = number
      protocol = string
    }))`,
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.try([
      Fn.lookup(Fn.element(list.value, index.value), "internal", "waaat"),
      Fn.timestamp(),
      Fn.uuid(),
    ]),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${try(lookup(element(var.list, var.index), \\"internal\\", \\"waaat\\"), timestamp(), uuid())}"
        }
      },
      "variable": {
        "index": {
          "type": "number"
        },
        "list": {
          "type": "list(object({\\n      internal = number\\n      external = number\\n      protocol = string\\n    }))"
        }
      }
    }"
  `);
});

test("function with varadic args", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "list(number)",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.merge([variable.value, [1, 2, 3]]),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${merge(var.test-var, [1, 2, 3])}"
        }
      },
      "variable": {
        "test-var": {
          "type": "list(number)"
        }
      }
    }"
  `);
});

test("complex example", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable1 = new TerraformVariable(stack, "test-var1", {
    type: "object({ key = number})",
  });
  const variable2 = new TerraformVariable(stack, "test-var2", {
    type: "object({ key = number})",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.cidrsubnet(
      Fn.lookup(Fn.merge([variable1.value, variable2.value]), "key", "default"),
      4,
      2
    ),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${cidrsubnet(lookup(merge(var.test-var1, var.test-var2), \\"key\\", \\"default\\"), 4, 2)}"
        }
      },
      "variable": {
        "test-var1": {
          "type": "object({ key = number})"
        },
        "test-var2": {
          "type": "object({ key = number})"
        }
      }
    }"
  `);
});

test("interpolation within string ", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: "number",
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.cidrsubnet(`172.16.0.0/${variable.value}`, 2, 3),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${cidrsubnet(\\"172.16.0.0/\${var.test-var}\\", 2, 3)}"
        }
      },
      "variable": {
        "test-var": {
          "type": "number"
        }
      }
    }"
  `);
});

test("functions with object inputs", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: `number`,
  });

  new TerraformOutput(stack, "test-output", {
    value: Fn.lookup(
      { var: variable.value, stat: 4, internal: true, yes: "no" },
      "internal",
      "waaat"
    ),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${lookup({\\"var\\" = var.test-var, \\"stat\\" = 4, \\"internal\\" = true, \\"yes\\" = \\"no\\"}, \\"internal\\", \\"waaat\\")}"
        }
      },
      "variable": {
        "test-var": {
          "type": "number"
        }
      }
    }"
  `);
});

test("quoted primitives in list", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.join(", ", ["world", "hello"]),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${join(\\", \\", [\\"world\\", \\"hello\\"])}"
        }
      }
    }"
  `);
});

test("quoted primitives, unquoted functions", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.join(", ", [Fn.join(" ", Fn.reverse(["world", "hello"]))]),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${join(\\", \\", [join(\\" \\", reverse([\\"world\\", \\"hello\\"]))])}"
        }
      }
    }"
  `);
});

test("nested objects and arrays as args", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  new TerraformOutput(stack, "test-output", {
    value: Fn.jsonencode({
      Statement: [
        {
          Action: "sts:AssumeRole",
          Effect: "Allow",
          Principal: { Service: "lambda.amazonaws.com" },
        },
      ],
      Version: "2012-10-17",
    }),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${jsonencode({\\"Statement\\" = [{\\"Action\\" = \\"sts:AssumeRole\\", \\"Effect\\" = \\"Allow\\", \\"Principal\\" = {\\"Service\\" = \\"lambda.amazonaws.com\\"}}], \\"Version\\" = \\"2012-10-17\\"})}"
        }
      }
    }"
  `);
});

test("terraform local", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const local = new TerraformLocal(stack, "list", ["world", "hello"]);

  new TerraformOutput(stack, "test-output", {
    value: Fn.reverse(local.asList),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "locals": {
        "list": [
          "world",
          "hello"
        ]
      },
      "output": {
        "test-output": {
          "value": "\${reverse(local.list)}"
        }
      }
    }"
  `);
});

test("undefined and null", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const local = new TerraformLocal(stack, "value", "hello world");

  new TerraformOutput(stack, "test-output", {
    value: Fn.coalesce([null, local.asString, undefined, 42, false]),
  });
  new TerraformOutput(stack, "json-object", {
    value: Fn.jsonencode({ a: "hello", b: 123, c: null, d: undefined }),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "locals": {
        "value": "hello world"
      },
      "output": {
        "json-object": {
          "value": "\${jsonencode({\\"a\\" = \\"hello\\", \\"b\\" = 123, \\"c\\" = null})}"
        },
        "test-output": {
          "value": "\${coalesce(local.value, 42, false)}"
        }
      }
    }"
  `);
});

describe("throws error on unescaped double quote string inputs", () => {
  const testsNotOk = ['"', '\\\\"', '\\\\\\\\"', '\\ \\\\"'];

  it.each(testsNotOk)("%s", (testNotOk) => {
    expect(() => {
      const app = Testing.app();
      const stack = new TerraformStack(app, "test");
      new TerraformOutput(stack, "test-output", {
        value: Fn.md5(testNotOk),
      });
      Testing.synth(stack);
    }).toThrowErrorMatchingSnapshot();
  });
});

describe("throws no error on valid escaped double quote string inputs", () => {
  const testsOk = [
    "",
    '\\"',
    '\\\\\\"',
    '\\ \\\\\\"',
    '\\ \\"',
    "\\abc\\def\\",
    "\\\\abc\\\\def\\\\",
  ];

  it.each(testsOk)("%s", (testOk) => {
    const app = Testing.app();
    const stack = new TerraformStack(app, "test");
    new TerraformOutput(stack, "test-output", {
      value: Fn.md5(testOk),
    });
    expect(Testing.synth(stack)).toMatchSnapshot();
  });
});

test("throws no error when wrapping unescaped double quotes in Fn.rawString", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");
  new TerraformOutput(stack, "test-output", {
    value: Fn.md5(Fn.rawString(`"`)),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "test-output": {
          "value": "\${md5(\\"\\\\\\"\\")}"
        }
      }
    }"
  `);
});

test("rawString escapes correctly", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");
  new TerraformLocal(stack, "test", {
    default: "abc",
    plain: Fn.rawString("abc"),
    infn: Fn.base64encode(Fn.rawString("abc")),
    quotes: Fn.rawString(`"`),
    doublequotes: Fn.rawString(`""`),
    template: Fn.rawString("${TEMPLATE}"),
  });

  const str = Testing.synth(stack);
  const json = JSON.parse(str);

  const bslsh = `\\`; // a single backslash
  expect(json.locals.test).toHaveProperty("default", "abc");
  expect(json.locals.test).toHaveProperty("plain", "abc");
  expect(json.locals.test).toHaveProperty("infn", '${base64encode("abc")}');
  expect(json.locals.test).toHaveProperty("quotes", `${bslsh}"`);
  expect(json.locals.test).toHaveProperty("doublequotes", `${bslsh}"${bslsh}"`);
  expect(json.locals.test).toHaveProperty("template", "$${TEMPLATE}");
});

test("tomap does not destroy incoming ref", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  expect(
    Testing.synthScope(
      (scope) =>
        new TerraformLocal(
          scope,
          "test",
          Fn.tomap(Token.asString(ref("test.instance.attr", stack)))
        )
    )
  ).toMatchInlineSnapshot(`
    "{
      "locals": {
        "test": "\${tomap(test.instance.attr)}"
      }
    }"
  `);
});

it("errors mentioning function name and argument", () => {
  expect(() =>
    Fn.replace("this value is ok", `this one " not`, "this is okay")
  ).toThrowErrorMatchingInlineSnapshot(
    `"Argument 1 of replace failed the validation: Error: 'this one " not' can not be used as value directly since it has unescaped double quotes in it. To safely use the value please use Fn.rawString on your string."`
  );
});

test("Property access using lookup and lookupNested functions", () => {
  const app = Testing.app();
  const stack = new TerraformStack(app, "test");

  const variable = new TerraformVariable(stack, "test-var", {
    type: `object({a = object({b = string}), z = string})`,
  });

  new TerraformOutput(stack, "lookup", {
    value: Fn.lookup(variable.value, "z", "defaultzzzz"),
  });
  new TerraformOutput(stack, "native-access", {
    value: Fn.lookup(variable.value, "z"),
  });
  new TerraformOutput(stack, "native-access-nested", {
    value: Fn.lookupNested(variable.value, ["a", "b"]),
  });

  expect(Testing.synth(stack)).toMatchInlineSnapshot(`
    "{
      "output": {
        "lookup": {
          "value": "\${lookup(var.test-var, \\"z\\", \\"defaultzzzz\\")}"
        },
        "native-access": {
          "value": "\${var.test-var.z}"
        },
        "native-access-nested": {
          "value": "\${var.test-var.a.b}"
        }
      },
      "variable": {
        "test-var": {
          "type": "object({a = object({b = string}), z = string})"
        }
      }
    }"
  `);
});
