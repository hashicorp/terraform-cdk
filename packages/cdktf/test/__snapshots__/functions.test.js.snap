// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws error on unescaped double quote string inputs " 1`] = `
"Argument 0 of md5 failed the validation: Error: '"' can not be used as value directly since it has unescaped double quotes in it.

To safely use the value, use Fn.rawString on your string like so:

Fn.rawString('"')

This is needed as CDKTF or Terraform will otherwise try to interpret these double quotes incorrectly.

To learn more about built in Terraform functions within CDKTF see: https://developer.hashicorp.com/terraform/cdktf/concepts/functions
. Please change your code to pass a valid value for this argument."
`;

exports[`throws error on unescaped double quote string inputs \\ \\\\" 1`] = `
"Argument 0 of md5 failed the validation: Error: '\\ \\\\"' can not be used as value directly since it has unescaped double quotes in it.

To safely use the value, use Fn.rawString on your string like so:

Fn.rawString('\\ \\\\"')

This is needed as CDKTF or Terraform will otherwise try to interpret these double quotes incorrectly.

To learn more about built in Terraform functions within CDKTF see: https://developer.hashicorp.com/terraform/cdktf/concepts/functions
. Please change your code to pass a valid value for this argument."
`;

exports[`throws error on unescaped double quote string inputs \\\\" 1`] = `
"Argument 0 of md5 failed the validation: Error: '\\\\"' can not be used as value directly since it has unescaped double quotes in it.

To safely use the value, use Fn.rawString on your string like so:

Fn.rawString('\\\\"')

This is needed as CDKTF or Terraform will otherwise try to interpret these double quotes incorrectly.

To learn more about built in Terraform functions within CDKTF see: https://developer.hashicorp.com/terraform/cdktf/concepts/functions
. Please change your code to pass a valid value for this argument."
`;

exports[`throws error on unescaped double quote string inputs \\\\\\\\" 1`] = `
"Argument 0 of md5 failed the validation: Error: '\\\\\\\\"' can not be used as value directly since it has unescaped double quotes in it.

To safely use the value, use Fn.rawString on your string like so:

Fn.rawString('\\\\\\\\"')

This is needed as CDKTF or Terraform will otherwise try to interpret these double quotes incorrectly.

To learn more about built in Terraform functions within CDKTF see: https://developer.hashicorp.com/terraform/cdktf/concepts/functions
. Please change your code to pass a valid value for this argument."
`;

exports[`throws no error on valid escaped double quote string inputs  1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\ \\" 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\ \\\\\\"\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\ \\\\\\" 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\ \\\\\\\\\\\\\\"\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\" 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\\\"\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\\\\\" 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\\\\\\\\\\\"\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\\\abc\\\\def\\\\ 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\\\\\abc\\\\\\\\def\\\\\\\\\\")}"
    }
  }
}"
`;

exports[`throws no error on valid escaped double quote string inputs \\abc\\def\\ 1`] = `
"{
  "output": {
    "test-output": {
      "value": "\${md5(\\"\\\\abc\\\\def\\\\\\")}"
    }
  }
}"
`;
